### PR TITLE
Associations

### PR DESCRIPTION
**Pull request by Peter1907**
Hello, dear code reviewer.
giphy

In this branch, I've implemented the following:

- Divided the code of the decorator file into separate files for each class as suggested in my last code review.
- Added new classes (book, classroom, rental).
- Added associations between classes as required.

Let me know if there is anything to be changed.
Thank you!

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 require './person'
 
 class Student < Person
+  attr_accessor :classroom
+
   def initialize(age, classroom, name = 'Unknown', parent_permission: true)
     super(age, name, parent_permission)
     @classroom = classroom
```
Comment (HAMMAS-SALEEM) [inline student.rb]: - [ ] When setting the classroom for a student it also adds it to the classrooms' students. 🤔 Kindly add the following method in your `Student` class 👍
```ruby
  def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
```

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+    rental.book = self
+  end
```
Comment (HAMMAS-SALEEM) [inline book.rb]: - [ ] Your add_rental method should look something like below ⬇️ to make many-to-many association 👍
```ruby
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
```

File: rental.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Rental
+  attr_accessor :date, :book, :person
+
+  def initialize(date, book, person)
+    @date = date
+    @book = book
+    @person = person
+  end
```
Comment (HAMMAS-SALEEM) [inline rental.rb]: - [ ] A rental belongs to a book and also belongs to a person: 🤔 So your initialize method should look something like below ⬇️
```ruby
def initialize(date, book, person)
    @book = book
    book.rentals << self
    @person = person
    person.rentals << self
  end
```

File: person.rb [hunk 4]
```diff
@@ -22,6 +23,11 @@ def can_use_services?
     of_age?
   end
 
+  def add_rental(rental)
+    @rentals << rental
+    rental.person = self
+  end
```
Comment (HAMMAS-SALEEM) [inline person.rb]: - [ ] Your add_rental method should look something like below to make a has_many association 👍
```ruby
  def add_rental(book, date)
    Rental.new(date, book, self)
  end
```

**Review by HAMMAS-SALEEM — CHANGES_REQUESTED** [review]
# Required Changes :recycle:
almost-there

## Hey @Peter1907 :wave:

## Highlights :1st_place_medal:

✔️ Descriptive PR Title
✔️ Descriptive PR Summary
✔️ Following Gitflow
✔️ Following some requirements

Good job on implementing some of the requirements, there are some issues that you still need to work on to go to the next project but you are almost there! :blush:

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Peter1907** [comment]
Thank you so much @HAMMAS-SALEEM :100: :+1:

**Review by Wes-Isaac — APPROVED** [review]
Hi @Peter1907, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

-  Descriptive PR Title 🥇.
-  Descriptive PR Summary 💯.
- Professional README 👍.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**